android/src/main/java/com/oney/WebRTCModule/PeerConnectionObserver.java

```diff
@@ -452,9 +359,124 @@ public void onSignalingChange(PeerConnection.SignalingState signalingState) {
+            WritableArray streams = Arguments.createArray();
+
+            for (MediaStream stream : mediaStreams) {
+                // Getting the streamReactTag 
+                String streamReactTag = null;
+                for (Map.Entry<String, MediaStream> e : remoteStreams.entrySet()) {
+                    if (e.getValue().equals(stream)) {
+                        streamReactTag = e.getKey();
+                        break;
+                    }
+                }
+                if (streamReactTag == null) {
+                    streamReactTag = UUID.randomUUID().toString();
+                    remoteStreams.put(streamReactTag, stream);
+                }
+                streams.pushMap(serializeStream(streamReactTag, stream));
+            }
+
+            params.putArray("streams", streams);
+            params.putMap("receiver", serializeReceiver(receiver));
+            params.putInt("transceiverOrder", getNextTransceiverId());
+            // Getting the transceiver object associated with the receiver for the event
+            List<RtpTransceiver> transceivers = peerConnection.getTransceivers();
+            for( RtpTransceiver transceiver : transceivers ) {
+                if(transceiver.getReceiver() != null && receiver.id().equals(transceiver.getReceiver().id())) {
+                    params.putMap("transceiver", serializeTransceiver(transceiver));
+                    break;
+                }
+            }
+            params.putInt("id", this.id);
+
+            webRTCModule.sendEvent("peerConnectionOnTrack", params);
+        });
+    }
+
+    /**
+     * Triggered when the signaling from SetRemoteDescription indicates that a transceiver
+     * will be receiving media from a remote endpoint. This is only called if UNIFIED_PLAN
+     * semantics are specified. The transceiver will be disposed automatically.
+     */
+    @Override
+    public void onTrack(final RtpTransceiver transceiver) {
+        Log.d(TAG, "onTrack");
+
+        ThreadUtils.runOnExecutor(() -> {
+            RtpReceiver receiver = transceiver.getReceiver();
+            MediaStreamTrack track = receiver.track();
+            if (track == null) {
+                return;
+            }
+            if(track.kind().equals(MediaStreamTrack.VIDEO_TRACK_KIND)){
+                videoTrackAdapters.addAdapter((VideoTrack) track);
+            }
+            remoteTracks.put(track.id(), track);
+
+            WritableMap params = Arguments.createMap();
+            WritableArray streams = Arguments.createArray();
+            
+            params.putArray("streams", streams);
+            params.putMap("receiver", serializeReceiver(receiver));
+            params.putMap("transceiver", serializeTransceiver(transceiver));
+            params.putInt("transceiverOrder", getNextTransceiverId());
+            params.putInt("id", this.id);
+
+            webRTCModule.sendEvent("peerConnectionOnTrack", params);
```

Comment:
Do we also need to update the SDP representations in the JS side? I think so?

Reply:
If you are referring to updating the transceivers, then that is already done as we are serialising the transceivers here as well. On the JS side, the objects themselves are being updated.

Reply:
No I mean the SDPs for pc.RemoteDrscriptipn.sdp and so on.

Reply:
I believe that would have happened prior to this event being invoked based on the description of OnTrack: 
```
    /**
     * Triggered when the signaling from SetRemoteDescription indicates that a transceiver
     * will be receiving media from a remote endpoint. This is only called if UNIFIED_PLAN
     * semantics are specified. The transceiver will be disposed automatically.
     */
```

Reply:
Maybe after addTrack / removeTrack ?

Reply:
if negotiation is necessary after those calls, the native side would fire up a negotiation needed event that propagates to the JS side, at which point subsequent calls to createOffer and setRemoteDescription should update the sdps, right?